cast/src/lib.rs

```diff
@@ -628,6 +629,103 @@ where
         Ok(receipt)
     }
 
+    pub async fn events(&self, hash: String, chain: Chain, api_key: String) -> eyre::Result<String> {
+        let receipt = self.provider.get_transaction_receipt(
+            H256::from_str(&hash)?
+        ).await?.unwrap();
+
+        // Address -> Vector of Events in the contract at that address
+        let mut cached_abis: HashMap<H160, Option<Vec<Event>>> = HashMap::new();
+        let mut return_string: String = String::new();
+
+        for log in receipt.logs {
+            // First, try to grab the ABI from the cache.
+            let abi: Option<Vec<Event>> = if let Some(cached_abi) = cached_abis.get(&log.address) {
+                cached_abi.clone()
+            
+            // If it's not there, go to Etherscan to look up the contract, and parse the interface into an Event ABI.
+            } else {
+                let interfaces = SimpleCast::generate_interface(InterfacePath::Etherscan {
+                    chain: chain,
+                    api_key: api_key.clone(),
+                    address: format!("{:#x}", log.address)
+                        .parse::<Address>()
+                        .wrap_err("Invalid address provided. Did you make a typo?")?,
+                }).await;
+
+                if let Ok(interfaces) = interfaces {
+                    let interface: Vec<Event> = interfaces
+                        .iter()
+                        .map(|iface| iface.source.to_string())
+                        .collect::<Vec<_>>()
+                        .join("\n")
+                        .split("\n")
+                        .map(|line| line.trim().to_string())
+                        .filter(|line| line.starts_with("event"))
+                        .map(|line| translate_event_string_to_event_type(&line).unwrap())
```

Comment:
you can use this instead or even `fn get_event()` located at `lib/`
`HumanReadableParser::parse_event(sig)`

Reply:
> Regarding your tradeoff, we handle that on traces by getting the signature from sam's db and use the following function:
> 
> https://github.com/foundry-rs/foundry/blob/be07dda496604d53493ec4ae9cd7b553a2d70a61/utils/src/lib.rs#L290-L292

@joshieDo Interesting. That logic seems to be that if all args are indexed, it adds it. Or if the number indexed is the same as addresses, it assumes addresses are indexed. Otherwise, it gives up.

That seems fine for traces where a "best guess" is fine, but for this function, it seems to me that making incorrect guesses isn't a worthwhile trade off to enable 4byte.

Thoughts?